connectd/onion_message.c

```diff
@@ -115,17 +115,39 @@ void handle_onion_message(struct daemon *daemon,
 {
 	struct pubkey blinding;
 	u8 *onion;
+	u64 msec;
+	struct timemono now = time_mono();
 
 	/* Ignore unless explicitly turned on. */
 	if (!feature_offered(daemon->our_features->bits[NODE_ANNOUNCE_FEATURE],
 			     OPT_ONION_MESSAGES))
 		return;
 
-	/* FIXME: ratelimit! */
+	/* Adjust tokens if they're entitled to more */
+	msec = time_to_msec(timemono_between(now, peer->onionmsg_last_incoming));
+	peer->onionmsg_incoming_tokens += msec;
+	if (peer->onionmsg_incoming_tokens > ONION_MSG_TOKENS_MAX)
+		peer->onionmsg_incoming_tokens = ONION_MSG_TOKENS_MAX;
```

Comment:
would it be worth flipping onionmsg_limit_warned false here? (so resetting after frequency has decreased to less than 4/sec)